src/bin/pg_autoctl/supervisor.c

```diff
@@ -529,6 +469,77 @@ supervisor_stop(Supervisor *supervisor)
 }
 
 
+/*
+ * If we have received a signal that instructs a shutdown, such as SIGTERM or
+ * SIGINT, then we need to do one of these things:
+ *
+ * - first time we receive the signal, begin a shutdown sequence for all
+ *   services and the main supervisor itself,
+ *
+ * - when receiving the signal again, if it's a SIGTERM continue the shutdown
+ *   sequence,
+ *
+ * - when receiving a SIGINT forward it to our services so as to finish as fast
+ *   as we can, and from then on always use SIGINT (to that end we use
+ *   supervisor->shutdownSignal)
+ *
+ * Sending SIGTERM and then later SIGINT if the process is still running is a
+ * classic way to handle service shutdown.
+ */
+static void
+supervisor_handle_signals(Supervisor *supervisor)
+{
+	bool shutdownNow = false;
+	int signal = asked_to_stop_fast ? SIGINT : SIGTERM;
+	const char *signalStr = asked_to_stop_fast ? "SIGINT" : "SIGTERM";
+
+	if (asked_to_stop || asked_to_stop_fast)
```

Comment:
nitpick: Inverting this check will allow for an early return. And the same can be done by adding an `else` case that the if/elseif chain that sets `shutdownNow`. That way the `shutdownNow` variable is not needed at all anymore.

Reply:
Early return makes it easier to read the code, done now.

That said we have more cases than meet the eyes and I think we still need the `shutdownNow` variable: if supervisor->shutdownSequenceInProgress and we received another SIGTERM (asked_to_stop_fast is false), then we don't want to signal all processes again.

Or maybe that's what you're saying, that we should signal our services each time we receive a signal, be it SIGINT or SIGTERM? Then your comment makes more sense, so that's what I have now pushed.